core/bloom.go

```diff
@@ -0,0 +1,358 @@
+	// This is just a property check for the bloom struct.
+	if value == "" {
+		return RESP_MINUS_1, errEmptyValue
+	}
+
+	// Get the indexes where bits are supposed to be set
+	indexes, err := b.opts.getIndexes(value)
+	if err != nil {
+		fmt.Println("error in getting indexes for value:", value, "err:", err)
+		return RESP_MINUS_1, errUnableToHash
+	}
+
+	// Set the bits and keep a count of already set ones
+	count := 0
+	for _, v := range indexes {
+		if isBitSet(b.bitset, int(v)) {
+			count++
+		} else {
+			setBit(b.bitset, int(v))
+		}
+	}
+
+	if count == len(indexes) {
+		// All the bits were already set, return 0 in that case.
+		return RESP_ZERO, nil
+	}
+
+	return RESP_ONE, nil
+}
+
+// exists checks if the given `value` exists in the filter or not.
+// It hashes the given value and checks if the bits are set or not in
+// the underlying bitset. Returns "-1" in case of errors, "0" if the
+// element surely does not exist in the filter, and "1" if the element
+// may or may not exist in the filter.
+func (b *Bloom) exists(value string) ([]byte, error) {
+	// We're sure that empty values will be handled upper functions itself.
+	// This is just a property check for the bloom struct.
+	if value == "" {
+		return RESP_MINUS_1, errEmptyValue
+	}
+
+	// Get the indexes where bits are supposed to be set
+	indexes, err := b.opts.getIndexes(value)
+	if err != nil {
+		fmt.Println("error in getting indexes for value:", value, "err:", err)
+		return RESP_MINUS_1, errUnableToHash
+	}
+
+	// Check if all the bits at given indexes are set or not
+	// Ideally if the element is present, we should find all set bits.
+	for _, v := range indexes {
+		if !isBitSet(b.bitset, int(v)) {
+			// Return with "0" as we found one non-set bit (which is enough to conclude)
+			return RESP_ZERO, nil
+		}
+	}
+
+	// We reached here, which means the element may exist in the filter. Return "1" now.
+	return RESP_ONE, nil
+}
+
+// getIndexes returns a list of indexes of the underlying bit array where
+// bits are supposed to be set (to 1). It uses the set hash function against
+// the given `value` and caps the index with the total number of bits.
+func (opts *BloomOpts) getIndexes(value string) ([]uint64, error) {
+	indexes := make([]uint64, len(opts.hashFns))
+
+	// Iterate through the hash functions and get indexes
+	for i := 0; i < len(opts.hashFns); i++ {
+		fn := opts.hashFns[i]
+		fn.Reset()
+
+		if _, err := fn.Write([]byte(value)); err != nil {
+			return nil, err
+		}
+
+		// Save the index capped by total number of bits in the underlying array
+		indexes[i] = fn.Sum64() % opts.bits
+	}
+
+	return indexes, nil
+}
+
+// evalBFInit evaluates the BFINIT command responsible for initializing a
+// new bloom filter and allocation it's relevant parameters based on given inputs.
+func evalBFInit(args []string) []byte {
+	if len(args) != 3 {
+		return Encode(fmt.Errorf("%w for 'BFINIT' command", errWrongArgs), false)
+	}
```

Comment:
Let's have the option to use a bloom filter without initializing. If done then we use default options to create one. This would reduce the load on the user to understand error rate and capacity.

Reply:
Makes sense. Added.